plugin/pkg/admission/security/podsecuritypolicy/admission.go

```diff
@@ -111,11 +112,59 @@ func (a *podSecurityPolicyPlugin) SetInternalKubeInformerFactory(f informers.Sha
 // 3.  Try to generate and validate a PSP with providers.  If we find one then admit the pod
 //     with the validated PSP.  If we don't find any reject the pod and give all errors from the
 //     failed attempts.
-func (c *podSecurityPolicyPlugin) Admit(a admission.Attributes) error {
+func (c *PodSecurityPolicyPlugin) Admit(a admission.Attributes) error {
 	if a.GetResource().GroupResource() != api.Resource("pods") {
 		return nil
 	}
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*api.Pod)
+	// if we can't convert then fail closed since we've already checked that this is supposed to be a pod object.
+	// this shouldn't normally happen during admission but could happen if an integrator passes a versioned
+	// pod object rather than an internal object.
+	if !ok {
+		return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+	}
+
+	// if this is an update, see if we are only updating the ownerRef/finalizers.  Garbage collection does this
+	// and we should allow it in general, since you had the power to update and the power to delete.
+	// The worst that happens is that you delete something, but you aren't controlling the privileged object itself
+	if a.GetOperation() == admission.Update && rbacregistry.IsOnlyMutatingGCFields(a.GetObject(), a.GetOldObject(), apiequality.Semantic) {
+		return nil
+	}
+
+	// compute the context. If the current security context is valid, this call won't change it.
+	allowMutation := a.GetOperation() == admission.Create // TODO(liggitt): allow spec mutation during initializing updates?
+	allowedPod, pspName, validationErrs, err := c.computeSecurityContext(a, pod, allowMutation)
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+	if allowedPod != nil {
+		*pod = *allowedPod
+		// annotate and accept the pod
+		glog.V(4).Infof("pod %s (generate: %s) in namespace %s validated against provider %s", pod.Name, pod.GenerateName, a.GetNamespace(), pspName)
+		if pod.ObjectMeta.Annotations == nil {
+			pod.ObjectMeta.Annotations = map[string]string{}
+		}
+		// set annotation to mark this as passed. Note, that the actual value is not important, the
+		// validating PSP might even change later-on. Also not that pspName can be the empty string
+		// if failOnNoPolicies is false.
+		// TODO: if failOnNoPolicies is toggled from false to true, we will never update the annotation anymore. Is this desired?
+		pod.ObjectMeta.Annotations[psputil.ValidatedPSPAnnotation] = pspName
```

Comment:
@deads2k We could skip the mutation phase for non-create operations. But this annotation forces us to keep it for the update case. It's essential for the tests right now to expose the selected policy. If we change the tests, we might get rid of this and simplify the mutation phase to be create-only. wdyt?

Reply:
> @deads2k We could skip the mutation phase for non-create operations. But this annotation forces us to keep it for the update case. It's essential for the tests right now to expose the selected policy. If we change the tests, we might get rid of this and simplify the mutation phase to be create-only. wdyt?

That sounds good as a future improvement.  I'd like to keep this initial pull tight.

Reply:
Hmm, I would say that this annotation is broken by this split. For it to be accurate, you would need to require that the subsequent validate step validates using the PSP named here.

It would actually simplify the model (and speed it up) if we could require that updates to the pod be valid under the original PSP, but I guess that's a significant breaking API change.

Reply:
The annotation is for testing and for information only, or in other words: it's broken by design. You cannot trust its value in any subsequent non-mutating request. Actually, given the semantics of PSPs to be highly implicit (via an existantial quantifier), I would prefer to have this annotation go away. Then we can restrict the mutation phase to CREATE requests only and do just validation after that for any other request.

Reply:
SGTM. Follow up?

Reply:
> SGTM. Follow up?

Yeah.  I see some value in the annotation to allow an ordering during the validation that will probably let us finish faster.  Not mutating at all during update seems like a great change though.